File: ios/RNFirebase/fabric/crashlytics/RNFirebaseCrashlytics.m
#import "RNFirebaseCrashlytics.h"

#if __has_include(<Crashlytics/Crashlytics/Crashlytics.h>)
    #define HAS_CRASHLYTICS 1
    #import <Crashlytics/Crashlytics/Crashlytics.h>
#else
    #if __has_include(<Crashlytics/Crashlytics.h>)
    #define HAS_CRASHLYTICS 2
    #import <Crashlytics/Crashlytics.h>
    #endif
#endif

#if __has_include(<Fabric/Fabric.h>)
    #import <Fabric/Fabric.h>
#endif

#ifdef HAS_CRASHLYTICS
    @implementation RNFirebaseCrashlytics
    RCT_EXPORT_MODULE();

    RCT_EXPORT_METHOD(crash) {
        [[Crashlytics sharedInstance] crash];
    }

    RCT_EXPORT_METHOD(log:(NSString *)message) {
        CLS_LOG(@"%@", message);
    }

    RCT_EXPORT_METHOD(recordError:(nonnull NSNumber *)code domain:(NSString *)domain) {
        NSError *error = [NSError errorWithDomain:domain code:[code integerValue] userInfo:nil];
        [CrashlyticsKit recordError:error];
    }

    RCT_EXPORT_METHOD(setBoolValue:(NSString *)key boolValue:(BOOL *)boolValue) {
        [CrashlyticsKit setBoolValue:boolValue forKey:key];
    }

    RCT_EXPORT_METHOD(setFloatValue:(NSString *)key floatValue:(nonnull NSNumber *)floatValue) {
        [CrashlyticsKit setFloatValue:[floatValue floatValue] forKey:key];
    }

    RCT_EXPORT_METHOD(setIntValue:(NSString *)key intValue:(nonnull NSNumber *)intValue) {
        [CrashlyticsKit setIntValue:[intValue integerValue] forKey:key];
    }

    RCT_EXPORT_METHOD(setStringValue:(NSString *)key stringValue:(NSString *)stringValue) {
        [CrashlyticsKit setObjectValue:stringValue forKey:key];
    }

    RCT_EXPORT_METHOD(setUserIdentifier:(NSString *)userId) {
        [CrashlyticsKit setUserIdentifier:userId];
    }

    RCT_EXPORT_METHOD(enableCrashlyticsCollection) {
#if __has_include(<Fabric/Fabric.h>)
        [Fabric with:@[[Crashlytics class]]];
#endif
    }

    @end
#else
    @implementation RNFirebaseCrashlytics
    @end
#endif

File: android/src/main/java/io/invertase/firebase/fabric/crashlytics/RNFirebaseCrashlytics.java
package io.invertase.firebase.fabric.crashlytics;

import android.util.Log;

import com.crashlytics.android.Crashlytics;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;

import java.util.ArrayList;

import io.fabric.sdk.android.Fabric;

public class RNFirebaseCrashlytics extends ReactContextBaseJavaModule {

  private static final String TAG = "RNFirebaseCrashlytics";

  public RNFirebaseCrashlytics(ReactApplicationContext reactContext) {
    super(reactContext);
    Log.d(TAG, "New instance");
  }

  @Override
  public String getName() {
    return TAG;
  }

  @ReactMethod
  public void crash() {
    Crashlytics
      .getInstance()
      .crash();
  }

  @ReactMethod
  public void log(final String message) {
    Crashlytics.log(message);
  }

  @ReactMethod
  public void recordError(final int code, final String domain) {
    Crashlytics.logException(new Exception(code + ": " + domain));
  }

  @ReactMethod
  public void recordCustomError(String name, String reason, ReadableArray frameArray) {
      ArrayList<StackTraceElement> stackList = new ArrayList<>(0);
      for (int i = 0; i < frameArray.size(); i++) {
        ReadableMap map = frameArray.getMap(i);
        ReadableMap additional = map.hasKey("additional") ? map.getMap("additional") : null;
        String functionName = map.hasKey("functionName") ? map.getString("functionName") : "Unknown Function";
        String className = map.hasKey("className") ? map.getString("className") : "Unknown Class";
        StackTraceElement stack = new StackTraceElement(
          className,
          functionName,
          map.getString("fileName"),
          map.hasKey("lineNumber") ? map.getInt("lineNumber") : -1
        );
        stackList.add(stack);

        if(additional != null){
          StackTraceElement s = new StackTraceElement(
            "Additional Parameters",
            additional.toString(),
            map.getString("fileName"),
            map.hasKey("lineNumber") ? map.getInt("lineNumber") : -1
          );
          stackList.add(s);
        }
      }
      StackTraceElement[] stackTrace =  new StackTraceElement[stackList.size()];
      Exception e = new Exception(name + "\n" + reason);
      stackTrace = stackList.toArray(stackTrace);
      e.setStackTrace(stackTrace);
      Crashlytics.logException(e);
  }

  @ReactMethod
  public void setBoolValue(final String key, final boolean boolValue) {
    Crashlytics.setBool(key, boolValue);
  }

  @ReactMethod
  public void setFloatValue(final String key, final float floatValue) {
    Crashlytics.setFloat(key, floatValue);
  }

  @ReactMethod
  public void setIntValue(final String key, final int intValue) {
    Crashlytics.setInt(key, intValue);
  }

  @ReactMethod
  public void setStringValue(final String key, final String stringValue) {
    Crashlytics.setString(key, stringValue);
  }

  @ReactMethod
  public void setUserIdentifier(String userId) {
    Crashlytics.setUserIdentifier(userId);
  }

  @ReactMethod
  public void setUserName(String userName) {
    Crashlytics.setUserName(userName);
  }

  @ReactMethod
  public void setUserEmail(String userEmail) {
    Crashlytics.setUserEmail(userEmail);
  }

  @ReactMethod
  public void enableCrashlyticsCollection() {
    Fabric.with(getReactApplicationContext(), new Crashlytics());
  }

}
